File: src/tool-server/tools/adapters/read.ts
/**
 * Read Tool Adapter - Read file contents
 */

import { join, isAbsolute, resolve } from "path";
import { registerTool, type ToolContext, type ToolResult } from "../registry";

const MAX_LINES = 2000;
const MAX_LINE_LENGTH = 2000;

async function readImpl(
  args: Record<string, unknown>,
  context: ToolContext
): Promise<ToolResult> {
  const filePath = args.file_path as string;
  const offset = (args.offset as number) ?? 0;
  const limit = (args.limit as number) ?? MAX_LINES;

  if (!filePath) {
    return {
      content: "Missing required parameter: file_path",
      status: "error",
    };
  }

  // Resolve path relative to session workspace
  const resolvedPath = isAbsolute(filePath)
    ? filePath
    : resolve(context.cwd, filePath);

  // Security: Check if path is within workspace
  const workspaceRoot = context.session.workspaceRoot;
  if (!resolvedPath.startsWith(workspaceRoot) && !isAbsolute(filePath)) {
    // Allow absolute paths for now (they might be system files)
    // In production, you might want to restrict this
  }

  try {
    const file = Bun.file(resolvedPath);

    if (!(await file.exists())) {
      return {
        content: `File not found: ${resolvedPath}`,
        status: "error",
      };
    }

    const content = await file.text();

    // Check for binary content (simple heuristic)
    if (content.includes("\x00")) {
      return {
        content: `Cannot read binary file: ${resolvedPath}`,
        status: "error",
      };
    }

    // Split into lines and apply offset/limit
    const lines = content.split("\n");
    const totalLines = lines.length;
    const startLine = Math.min(offset, totalLines);
    const endLine = Math.min(startLine + limit, totalLines);
    const selectedLines = lines.slice(startLine, endLine);

    // Format with line numbers and truncate long lines
    const formattedLines = selectedLines.map((line, index) => {
      const lineNum = startLine + index + 1;
      const truncatedLine =
        line.length > MAX_LINE_LENGTH
          ? line.slice(0, MAX_LINE_LENGTH) + "..."
          : line;
      // Format: "   123→content" (right-aligned line number)
      const lineNumStr = String(lineNum).padStart(6, " ");
      return `${lineNumStr}→${truncatedLine}`;
    });

    let output = formattedLines.join("\n");

    // Add truncation notice if needed
    if (endLine < totalLines) {
      output += `\n\n[Showing lines ${startLine + 1}-${endLine} of ${totalLines}. Use offset/limit to see more.]`;
    }

    if (context.onStdout) {
      context.onStdout(output);
    }

    return {
      content: output || "(empty file)",
      status: "success",
    };
  } catch (error) {
    const message = error instanceof Error ? error.message : "Unknown error";
    return {
      content: `Failed to read file: ${message}`,
      status: "error",
    };
  }
}

// Register the Read tool
registerTool({
  name: "Read",
  description: "Read file contents with optional line offset and limit",
  implementation: readImpl,
  schema: {
    type: "object",
    properties: {
      file_path: {
        type: "string",
        description: "Path to the file to read (absolute or relative to workspace)",
      },
      offset: {
        type: "number",
        description: "Line number to start reading from (0-indexed, default: 0)",
      },
      limit: {
        type: "number",
        description: `Number of lines to read (default: ${MAX_LINES})`,
      },
    },
    required: ["file_path"],
  },
});

File: gosdk/tools/read.go
package tools

import (
	"context"
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"unicode/utf8"
)

// ReadTool reads files from the local filesystem.
type ReadTool struct {
	BaseTool
	workDir string
}

// ReadArgs represents the arguments for the Read tool.
type ReadArgs struct {
	FilePath string `json:"file_path"`
	Offset   int    `json:"offset,omitempty"` // Line number to start from (0-based)
	Limit    int    `json:"limit,omitempty"`  // Number of lines to read
}

// NewReadTool creates a new Read tool instance.
func NewReadTool(workDir string) *ReadTool {
	return &ReadTool{
		BaseTool: BaseTool{
			name: "Read",
			description: `Reads a file from the local filesystem. You can access any file directly by using this tool.
Assume this tool is able to read all files on the machine. If the User provides a path to a file assume that path is valid. It is okay to read a file that does not exist; an error will be returned.

Usage:
- The file_path parameter must be an absolute path, not a relative path
- By default, it reads up to 2000 lines starting from the beginning of the file
- You can optionally specify a line offset and limit (especially handy for long files), but it's recommended to read the whole file by not providing these parameters
- Any lines longer than 2000 characters will be truncated
- Results are returned using cat -n format, with line numbers starting at 1
- This tool can only read files, not directories. To read a directory, use the ls command via Bash.
- You can call multiple tools in a single response. It is always better to speculatively read multiple potentially useful files in parallel.
- If you read a file that exists but has empty contents you will receive a system reminder warning in place of file contents.`,
		},
		workDir: workDir,
	}
}

// Call executes the read operation.
func (t *ReadTool) Call(ctx context.Context, input string) (string, error) {
	if err := validateContext(ctx); err != nil {
		return ErrorResult(err), nil
	}

	var args ReadArgs
	if err := json.Unmarshal([]byte(input), &args); err != nil {
		return ErrorResult(fmt.Errorf("invalid input: %w", err)), nil
	}

	if args.FilePath == "" {
		return ErrorResult(fmt.Errorf("file_path is required")), nil
	}

	// Resolve path
	resolvedPath := args.FilePath
	if !filepath.IsAbs(resolvedPath) {
		resolvedPath = filepath.Join(t.workDir, resolvedPath)
	}

	// Check if path exists and is not a directory
	stat, err := os.Stat(resolvedPath)
	if err != nil {
		if os.IsNotExist(err) {
			return ErrorResult(fmt.Errorf("file does not exist. Attempted path: %s. Current working directory: %s", resolvedPath, t.workDir)), nil
		}
		if os.IsPermission(err) {
			return ErrorResult(fmt.Errorf("permission denied: %s", resolvedPath)), nil
		}
		return ErrorResult(err), nil
	}

	if stat.IsDir() {
		return ErrorResult(fmt.Errorf("path is a directory, not a file: %s", resolvedPath)), nil
	}

	// Check file size (10MB max)
	maxSize := int64(10 * 1024 * 1024)
	if stat.Size() > maxSize {
		return ErrorResult(fmt.Errorf("file too large: %d bytes (max %d bytes)", stat.Size(), maxSize)), nil
	}

	// Check if binary file
	if isBinary, err := isBinaryFile(resolvedPath); err != nil {
		return ErrorResult(err), nil
	} else if isBinary {
		return ErrorResult(fmt.Errorf("cannot read binary file: %s", resolvedPath)), nil
	}

	// Read file content
	content, err := os.ReadFile(resolvedPath)
	if err != nil {
		return ErrorResult(err), nil
	}

	// Check for empty file
	if len(strings.TrimSpace(string(content))) == 0 {
		return SuccessResult(fmt.Sprintf("<system-reminder>\nThe file %s exists but has empty contents.\n</system-reminder>", resolvedPath)), nil
	}

	// Format with line numbers
	formatted := formatWithLineNumbers(string(content), args.Offset, args.Limit)
	return SuccessResult(formatted), nil
}

// isBinaryFile checks if a file appears to be binary.
func isBinaryFile(path string) (bool, error) {
	file, err := os.Open(path)
	if err != nil {
		return false, err
	}
	defer file.Close()

	// Read up to 8KB
	buf := make([]byte, 8192)
	n, err := file.Read(buf)
	if err != nil && n == 0 {
		return false, nil // Empty file is not binary
	}
	buf = buf[:n]

	// Check for null bytes (definite binary indicator)
	for _, b := range buf {
		if b == 0 {
			return true, nil
		}
	}

	// Check if valid UTF-8
	if !utf8.Valid(buf) {
		return true, nil
	}

	// Count control characters (excluding whitespace)
	controlCount := 0
	for _, b := range buf {
		if b < 9 || (b > 13 && b < 32) {
			controlCount++
		}
	}

	// If more than 30% control characters, consider binary
	return float64(controlCount)/float64(len(buf)) > 0.3, nil
}

// formatWithLineNumbers formats content with line numbers like cat -n.
func formatWithLineNumbers(content string, offset, limit int) string {
	lines := strings.Split(content, "\n")
	originalLineCount := len(lines)

	// Apply default limit if not specified
	effectiveLimit := limit
	if effectiveLimit <= 0 {
		effectiveLimit = Limits.ReadMaxLines
	}

	startLine := offset
	if startLine < 0 {
		startLine = 0
	}
	if startLine > len(lines) {
		startLine = len(lines)
	}

	endLine := startLine + effectiveLimit
	if endLine > len(lines) {
		endLine = len(lines)
	}

	selectedLines := lines[startLine:endLine]

	// Calculate padding for line numbers
	maxLineNum := startLine + len(selectedLines)
	padding := len(fmt.Sprintf("%d", maxLineNum))
	if padding < 1 {
		padding = 1
	}

	var result strings.Builder
	var truncatedLines bool

	for i, line := range selectedLines {
		lineNum := startLine + i + 1 // 1-based line numbers

		// Truncate long lines
		if len(line) > Limits.ReadMaxCharsPerLine {
			line = line[:Limits.ReadMaxCharsPerLine] + "... [line truncated]"
			truncatedLines = true
		}

		result.WriteString(fmt.Sprintf("%*d→%s", padding, lineNum, line))
		if i < len(selectedLines)-1 {
			result.WriteString("\n")
		}
	}

	// Add truncation notices
	var notices []string
	wasTruncatedByLineCount := endLine < originalLineCount

	if wasTruncatedByLineCount && limit <= 0 {
		notices = append(notices, fmt.Sprintf(
			"\n\n[File truncated: showing lines %d-%d of %d total lines. Use offset and limit parameters to read other sections.]",
			startLine+1, endLine, originalLineCount,
		))
	}

	if truncatedLines {
		notices = append(notices, fmt.Sprintf(
			"\n\n[Some lines exceeded %d characters and were truncated.]",
			Limits.ReadMaxCharsPerLine,
		))
	}

	for _, notice := range notices {
		result.WriteString(notice)
	}

	return result.String()
}
